OpInfo: Add natural syntax for SampleInput creation

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85890
* __->__ #85723

Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

peterbell10 committed:
OpInfo: Add helper function to simplify SampleInput creation

Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
the `make_sample_input` helper which takes its function arguments
and returns the corresponding `SampleInput` object.

This leads to a very natural syntax where you just call
`make_sample_input` as if it were the function being tested.
So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
make_sample_input(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -262,6 +262,24 @@ def to_noncontiguous(t):
 
 
+def make_sample_input(
+    input: Union[torch.Tensor, Sequence[torch.Tensor]], *args, **kwargs
```
Input isn't always a tensor or sequence of tensors anymore, which puts the code in an awkward place as its primacy doesn't make so much sense anymore. For now I think we could just model this as `Any`. 

I think adding an ability like this is very natural and I was thinking along similar lines. Do you think instead of adding this function we can change the constructor to `SampleInput` so developers can just write:

```
SampleInput(a, b, out=out)
```

just like calling the function? I understand that the SampleInput constructor accepts kwargs controlling its behavior, but I don't think these kwargs overlap with operator parameters?

```
output_process_fn_grad=lambda x: x,
broadcasts_input=False,
name="",
```

I would also like to kill the `broadcasts_input` and `name` params, but that's a future discussion. Basically I'm curious if we can update the signature for the SampleInput constructor to be:

```
(arg, *args, 
output_process_fn_grad=lambda x: x,
broadcasts_input=False,
name="",
**kwargs)
```

?

peterbell10 replied:
I'm not a big fan of this because the keyword arguments meant for the operator and those meant as metadata aren't clearly separated when reading the code. The assumption of no name collisions is also something that might come back to bite use later.

How about this syntax though:
```
SampleInput(a, b, out=out).output_process_fn_grad(fn)
```

The arguments are clearly separated from the metadata and 99% of the time you just call `SampleInput` with nothing extra.

peterbell10 replied:
I've updated the PR to allow `SampleInput` to be constructed with var-args and var-kwargs in a backwards-compatible way. It occurred to me that any edge cases that don't work can just fall back to the existing syntax, so it doesn't matter too much.

peterbell10 committed:
Update on "OpInfo: Add natural syntax for SampleInput creation"


Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add natural syntax for SampleInput creation"


Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -126,20 +126,44 @@ class SampleInput(object):
+        if kwargs is None:
+            self.kwargs = var_kwargs
+        else:
+            assert (
```
Style suggestion:

```
if args is not None or kwargs is not None:
  assert (len(var_args) == 0 and len(var_kwargs) == 0), "A SampleInput can be constructed "naturally" with *args and **kwargs or by explicitly setting the "args" and "kwargs" paremets, but the two methods of construction cannot be mixed!"

self.args = args if args is not None else var_args
self.kwargs = kwargs if kwargs is not None else var_kwargs
```

So this will still allow developers to set the SampleInput-specific `output_process_fn_grad`, `broadcasts_input`, and `name` parameters inline with function parameters? Per your comment, why not extend this check to disallow those being set if `**kwargs` has values? A message could direct the developer to set them using the fluent helpers you've suggested:

```
# throws an error
SampleInput(a, b, weights=c, output_process_fn=foo)

# works
SampleInput(a, b, weights=c).output_process_fn(foo)
```

cc @pmeier @lezcano if they'd like to help us bikeshed on these UX details

pmeier replied:
+1 on Mikes proposal. I prefer the fluent style, because this makes the `SampleInput` more clear cut.

We could even warn that the old style of passing everything to the constructor is deprecated, but I'm guessing most people will not look at the warnings anyway? If we change most `OpInfo`'s to this style sooner or later most people will start to use it since all references do.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -148,43 +148,13 @@ def sample_inputs_slice(op_info, device, dtype, requires_grad, **kwargs):
-            'end': -2,
-        }
-    )
+    yield SampleInput(make_input(20, 30, 40), dim=1, start=1, end=-2)
```
A million times better

mruberry commented on test/test_testing.py:
```diff
@@ -1814,6 +1815,35 @@ def test_no_warning_on_import(self) -> None:
+            s = SampleInput(a, args=(b, c), d=d, e=e)
+
+        with self.assertRaises(AssertionError):
+            s = SampleInput(a, b, c=c, kwargs=dict(d=d, e=e))
```
I would add cases for the SampleInput-specific params with whatever behavior is expected, too

mruberry commented on test/test_testing.py:
```diff
@@ -1814,6 +1815,35 @@ def test_no_warning_on_import(self) -> None:
             cwd=os.path.dirname(os.path.realpath(__file__)),).decode("utf-8")
         self.assertEquals(out, "")
 
+class TestOpInfos(TestCase):
```
Nice test addition

mruberry reviewed (approved):
This is a really nice UX improvement and I think it's fine as-is, but see inline suggestions for how we may want to tweak this

pmeier reviewed (commented):
Only additional nits, but otherwise LGTM. Thanks Peter!

peterbell10 committed:
Update on "OpInfo: Add natural syntax for SampleInput creation"


Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -126,22 +126,47 @@ class SampleInput(object):
+        # Allow calling either as SampleInput(input, args=args, kwargs=kwargs), or as
+        # SampleInput(input, *args, **kwargs) but not to mix the two forms
+        if args is not None or kwargs is not None:
+            assert (
+                not var_args and not var_kwargs
+            ), """
+            A SampleInput can be constructed "naturally" with *args and **kwargs or by
+            explicitly setting the "args" and "kwargs" paremeters, but the two
+            methods of construction cannot be mixed!"""
```
One more style nit: using `assert` with longer messages is generally a pain. Your example leads to a lot of whitespace in front of the message:

```py
class Foo:
    def __init__(self):
        assert False, """
        A SampleInput can be constructed "naturally" with *args and **kwargs or by
        explicitly setting the "args" and "kwargs" paremeters, but the two
        methods of construction cannot be mixed!"""


Foo()
```

```
AssertionError: 
        A SampleInput can be constructed "naturally" with *args and **kwargs or by
        explicitly setting the "args" and "kwargs" paremeters, but the two
        methods of construction cannot be mixed!
```

There are several options here in increasing order of my preference:

1. Left-align the multi-line string
    ```suggestion
                assert (
                    not var_args and not var_kwargs
                ), """
    A SampleInput can be constructed "naturally" with *args and **kwargs or by
    explicitly setting the "args" and "kwargs" paremeters, but the two
    methods of construction cannot be mixed!"""
    ```
2. Use parentheses to wrap multiple single strings rather than one multi-line string
    ```suggestion
                assert (
                    not var_args and not var_kwargs
                ), (
                "A SampleInput can be constructed 'naturally' with * args and ** kwargs or by "
                "explicitly setting the 'args' and 'kwargs' paremeters, but the two "
                "methods of construction cannot be mixed!"
            )
    ```
3. Drop the the `assert condition, message` idiom in favor of `if condition: raise AssertionError(message)`
    ```suggestion
                if not var_args and not var_kwargs:
                    raise AssertionError(
                        "A SampleInput can be constructed 'naturally' with * args and ** kwargs or by "
                        "explicitly setting the 'args' and 'kwargs' paremeters, but the two "
                        "methods of construction cannot be mixed!"
                    )
    ```

peterbell10 committed:
Update on "OpInfo: Add natural syntax for SampleInput creation"


Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add natural syntax for SampleInput creation"


Most SampleInput objects currently have no additional metadata,
meaning they have a 1:1 mapping with a normal function call. This adds
var arg forms of the `SampleInput` constructor such that you can just
call the `SampleInput` constructor as you would call the operator.

So, for example
```python
SampleInput(make_arg(shape), args=(2, 3), kwargs=dict(alpha=4))
```
becomes
```python
SampleInput(make_arg(shape), 2, 3, alpha=4)
```

[ghstack-poisoned]
